samples/nrf9160/modem_shell/src/utils/mosh_print.c

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <assert.h>
+
+#include <zephyr.h>
+#include <posix/time.h>
+#include <shell/shell.h>
+#include <shell/shell_uart.h>
+
+#include "mosh_print.h"
+
+const struct shell *shell_global;
+
+/** Configuration on whether timestamps are added to the print. */
+bool mosh_print_timestamp_use;
+
+/** Buffer used for printing the timestamp. */
+static char timestamp_str[30];
+/** Buffer used for generating the text to be printed. */
+static char mosh_print_buf[CONFIG_MOSH_PRINT_BUFFER_SIZE];
+/** Mutex for protecting mosh_print_buf */
+K_MUTEX_DEFINE(mosh_print_buf_mutex);
+
+static const char *create_timestamp_string(void)
+{
+	uint32_t year;
+	uint32_t month;
+	uint32_t day;
+
+	uint32_t hours;
+	uint32_t mins;
+	uint32_t secs;
+	uint32_t msec;
+
+	struct timespec tp;
+	struct tm ltm = { 0 };
+
+	clock_gettime(CLOCK_REALTIME, &tp);
+	gmtime_r(&tp.tv_sec, &ltm);
+
+	msec = tp.tv_nsec / 1000000;
+	secs = ltm.tm_sec;
+	mins = ltm.tm_min;
+	hours = ltm.tm_hour;
+	day = ltm.tm_mday;
+	month = ltm.tm_mon;
+	year = 1900 + ltm.tm_year;
+
+	sprintf(timestamp_str,
+		"[%04d-%02d-%02d %02d:%02d:%02d.%03d] ",
+		year, month, day, hours, mins, secs, msec);
+
+	return timestamp_str;
+}
+
+void mosh_fprintf(enum mosh_print_level print_level, const char *fmt, ...)
+{
+	va_list args;
+	int chars = 0;
+	const char *timestamp;
+
+	k_mutex_lock(&mosh_print_buf_mutex, K_FOREVER);
+
+	if (shell_global == NULL) {
+		shell_global = shell_backend_uart_get_ptr();
+		if (shell_global == NULL) {
+			printk("STILL CANNOT PRINT BECAUSE CANNOT OBTAIN SHELL BACKEND!!!\n");
```

Comment:
Does this have to be all uppercase? This would be printed for each call, so I think less shouting should be enough. :)

Reply:
This should never be printed. The reason I made it all upper case is that because it's not shell_error etc which would have colors to highlight the output, I wanted to make it clearly visible. If anybody sees this, we should investigate why. If it turns out this can be seen in normal scenarios, then we need to see if we need to remove the entire line. I had originally the same text on line 72 but now that line is hit when the first mosh_print call occurs.
So, I can chance this to normal text if required but I'm afraid that'll just hide a problem.

Reply:
I've modified this to:
```
			printk("mosh_fprintf(): CANNOT OBTAIN SHELL BACKEND. "
			       "THIS IS FATAL INTERNAL ERROR.\n");
```

Reply:
👍 That's more informative.